ChopChop/Models/Recipe/Recipe/Recipe.swift

```diff
@@ -0,0 +1,210 @@
+import Foundation
+import GRDB
+
+/// Note there is no relationship between steps and ingredients after parsing stage
+class Recipe: FetchableRecord {
+    var id: Int64?
+    private(set) var name: String
+    private(set) var servings: Double
+    var recipeCategoryId: Int64?
+    private(set) var difficulty: Difficulty?
+    private(set) var steps: [RecipeStep]
+    private(set) var ingredients: [RecipeIngredient]
+
+    init(name: String, servings: Double = 1, difficulty: Difficulty? = nil,
+         steps: [RecipeStep] = [], ingredients: [RecipeIngredient] = []) throws {
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+
+        guard servings > 0 else {
+            throw RecipeError.invalidServings
+        }
+        self.servings = servings
+        self.difficulty = difficulty
+        self.steps = steps
+        self.ingredients = ingredients
+        assert(checkRepresentation())
+    }
+
+    func updateName(_ name: String) throws {
+        assert(checkRepresentation())
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+        assert(checkRepresentation())
+    }
+
+    func updateDifficulty(_ difficulty: Difficulty) {
+        assert(checkRepresentation())
+        self.difficulty = difficulty
+        assert(checkRepresentation())
+    }
+
+    // step related functions
+    func addStep(content: String) throws {
+        assert(checkRepresentation())
+        steps.append(try RecipeStep(content: content))
+        assert(checkRepresentation())
+    }
+
+    func removeStep(_ removedStep: RecipeStep) {
+        assert(checkRepresentation())
+        guard (steps.contains { $0 == removedStep }) else {
+            return
+        }
+
+        steps.removeAll { $0 == removedStep }
+        assert(checkRepresentation())
+    }
+
+    func reorderStep(movedStep: RecipeStep, isSwapUp: Bool) throws {
```

Comment:
Instead of swapping neighbouring steps, maybe it will be more useful to be able to swap two steps at any two indexes or slot an existing step into a new index

Reply:
For adding, what I'm thinking is that the user will only be able to add a step at the end of the list, then will use reorder to get it to the right index.

Tbh I don' t think there will be any circumstance where 2 non-neighbouring steps will swap positions (if there r pls lmk). I guess I could do `reorderStep(movedStep: RecipeStep, index: int)`. The UI will probably just be up and down arrows for each step. And how it would be used is that the RecipeViewModel will find the index of the movedStep and +1 if pressed the down button and -1 if pressed the up button